feat(#173): add script hash history

jakubmkowalski opened the pull request:
This pull request introduces functionality for retrieving script history, including both confirmed and unconfirmed transactions, for a given script hash. It includes validation, API integration, and extensive test coverage to ensure the correctness of the implementation. The changes span across fixtures, service methods, and data structures, enabling robust handling of script history queries.

### Script History Retrieval Functionality:
* **Validation of Script Hashes**: Added a `validateScriptHash` function in `pkg/services/internal/whatsonchain/script_history.go` to ensure script hashes meet length and format requirements.
* **API Integration**: Implemented methods `getConfirmedScriptHistory` and `getUnconfirmedScriptHistory` in `pkg/services/internal/whatsonchain/script_history.go` to interact with the WhatsOnChain API and retrieve script history. Combined results are handled by the `GetScriptHistory` method.
* **Service Exposure**: Added `ScriptHistory` method to `WalletServices` in `pkg/services/services.go` for external access to script history retrieval.

### Fixture Enhancements for Testing:
* **Fixture Methods**: Extended `WhatsOnChainFixture` in `pkg/internal/testabilities/testservices/fixture_woc.go` with methods for mocking script history responses, including error handling and pagination. [[1]](diffhunk://#diff-0a498ce2b803514ad0826deb0f2d90e0d7664b8166232bf949179dda7f93cac9R37-R43) [[2]](diffhunk://#diff-0a498ce2b803514ad0826deb0f2d90e0d7664b8166232bf949179dda7f93cac9R357-R625)
* **Script History Data Builder**: Introduced `ScriptHistoryDataBuilder` to easily construct mock script history data for testing.

### Data Structure Updates:
* **Script History Models**: Added new types in `pkg/wdk/services_script_hash_history.go` to represent script history items, results, and options for querying confirmed history. Includes validation for query options.

### Test Coverage:
* **Unit Tests**: Added comprehensive tests in `pkg/services/services_get_script_history_test.go` to validate script history retrieval, including builder pattern usage, query fixture methods, validation errors, pagination, and large history scenarios.

These changes collectively enhance the ability to query, validate, mock, and test script history functionality within the application.## Description of Changes

Provide a brief description of the changes you've made.

## Linked Issues / Tickets

Closes #173 

## Testing Procedure

Describe the tests you've added or any testing steps you've taken.

- [ ] I have added new unit tests
- [ ] All tests pass locally
- [ ] I have tested manually in my local environment

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have run the linter

jakubmkowalski committed:
feat: implement script history retrieval for confirmed and unconfirmed transactions

jakubmkowalski committed:
feat: add script history retrieval tests with builder and query fixtures

jakubmkowalski committed:
feat: update package references from 4chain-ag to bsv-blockchain

jakubmkowalski committed:
feat: add error wrapping for WalletServices method GetScriptHistory

jakubmkowalski committed:
fix: correct function comment for GetScriptHistory in WalletServices

jakubmkowalski committed:
fix: remove debug print

jakubmkowalski commented on pkg/internal/testabilities/testservices/fixture_woc.go:
```diff
@@ -345,3 +354,272 @@ func computeTxID(rawTx []byte) string {
+		}
+	}
+
+	fmt.Println(url)
```
removed debug print

jakubmkowalski committed:
Merge branch 'main' into feat/add_script_hash_history

jakubmkowalski committed:
fix: remove unused structs

mgosek-4chain commented on pkg/internal/testabilities/testservices/fixture_woc.go:
```diff
@@ -345,3 +354,270 @@ func computeTxID(rawTx []byte) string {
+			httpmock.NewStringResponder(status, ""),
+		)
+	}
+}
```
Perhaps it would be better to split this into two separate methods? WDYT? 🤔

jakubmkowalski replied:
separated

mgosek-4chain commented on pkg/internal/testabilities/testservices/fixture_woc.go:
```diff
@@ -345,3 +354,270 @@ func computeTxID(rawTx []byte) string {
+			url,
+			httpmock.NewStringResponder(status, ""),
+		)
+	}
```
Same here :)

jakubmkowalski replied:
separated

mgosek-4chain commented on pkg/internal/testabilities/testservices/fixture_woc.go:
```diff
@@ -345,3 +354,270 @@ func computeTxID(rawTx []byte) string {
+	url := fmt.Sprintf("https://api.whatsonchain.com/v1/bsv/%s/script/%s/confirmed/history", q.fixture.network, q.scriptHash)
+
+	if opts != nil {
+		params := make([]string, 0)
```
`params := make([]string, 0, 4)` 🤔 

This reuses the same underlying array. Once the capacity is exceeded (e.g. 4), a new allocation will occur.  

4 because it's equal to number of the options (if statements).

jakubmkowalski replied:
this is now removed

mgosek-4chain commented on pkg/services/internal/whatsonchain/script_history.go:
```diff
@@ -0,0 +1,144 @@
+		if opts.NextPageToken != nil && *opts.NextPageToken != "" {
+			req.SetQueryParam("token", *opts.NextPageToken)
+		}
+	}
```
It might be clearer and more maintainable when you extend the `wdk.GetConfirmedScriptHistoryOpts` to have a method that returns a query string.

jakubmkowalski replied:
opts removed

mgosek-4chain commented on pkg/services/internal/whatsonchain/script_history_test.go:
```diff
@@ -0,0 +1,425 @@
+	s.registerUnconfirmedHistoryMock(scriptHash, http.StatusOK, unconfirmedResponse)
+
+	// when
+	result, err := s.woc.GetScriptHistory(context.Background(), scriptHash, nil)
```
`t.Context()`

jakubmkowalski replied:
changed

mgosek-4chain commented on pkg/services/internal/whatsonchain/script_history_test.go:
```diff
@@ -0,0 +1,425 @@
+			url,
+			httpmock.NewStringResponder(status, ""),
+		)
+	}
```
I would recommend to split this part into two separate methods as I suggested earlier.

jakubmkowalski replied:
splited

dorzepowski commented on pkg/services/services.go:
```diff
@@ -199,3 +199,12 @@ func (s *WalletServices) HashToHeader(hash string) (*wdk.ChainBlockHeader, error
 }
+
+// GetScriptHistory retrieves both confirmed and unconfirmed transaction history for a script hash
+func (ws *WalletServices) GetScriptHistory(ctx context.Context, scriptHash string, opts *wdk.GetConfirmedScriptHistoryOpts) (*wdk.ScriptHistoryResult, error) {
```
Although I like the idea of having possibility to pass some additional arguments to stear how this method is working, BUT:
1. The ts version doesn't have them, 
2. they should rather be optional - replaced with opts pattern at this level (the services could accept opts struct)
3. (minor) I'm not sure if they doesn't make the service implementation interface less flexible - because it they provides NextPageToken which is handled by WoC but could be done differently in other services
4. You didn't prepare any test that would show how NextPageToken could be used in a real world scenario

To wrap up:
Either change the opts to follow opts pattern - you can find [example here](https://github.com/bsv-blockchain/go-wallet-toolbox/blob/a125b2319bced8770ea6b01a8e2bf9ee5397068c/pkg/internal/testabilities/testservices/fixture_woc.go#L53)
or remove them

jakubmkowalski replied:
opts removed

dorzepowski commented on pkg/services/services.go:
```diff
@@ -199,3 +199,12 @@ func (s *WalletServices) HashToHeader(hash string) (*wdk.ChainBlockHeader, error
+
+// GetScriptHistory retrieves both confirmed and unconfirmed transaction history for a script hash
+func (ws *WalletServices) GetScriptHistory(ctx context.Context, scriptHash string, opts *wdk.GetConfirmedScriptHistoryOpts) (*wdk.ScriptHistoryResult, error) {
+	result, err := ws.whatsonchain.GetScriptHistory(ctx, scriptHash, opts)
```
Although we have only single implementation, but you should use servicequeue and OneByOne method of it, to make it extendable and configurable in the future.

See rawTx for example

jakubmkowalski replied:
now we use queue

dorzepowski commented on pkg/services/services.go:
```diff
@@ -199,3 +199,12 @@ func (s *WalletServices) HashToHeader(hash string) (*wdk.ChainBlockHeader, error
 	panic("Not implemented yet")
 }
+
+// GetScriptHistory retrieves both confirmed and unconfirmed transaction history for a script hash
```
I believe that the name in TS is GetScriptHashHistory and we should follow they naming

jakubmkowalski replied:
changed

dorzepowski commented on pkg/services/services_get_script_history_test.go:
```diff
@@ -0,0 +1,217 @@
+)
+
+// ScriptHistoryFixtureTestSuite demonstrates usage of the new fixture methods
+type ScriptHistoryFixtureTestSuite struct {
```
Sorry, but we're trying to avoid using the Suite's

The only place where we're using it is when we have several implementations that we want to check against the same set of tests.
And currently only the tests for wallet looks like the place for using the Suite (because there wasn't better solution).

And actually because of suite this test is made in different convention which can make others harder to reason about it.

jakubmkowalski replied:
removed

dorzepowski commented on pkg/services/services_get_script_history_test.go:
```diff
@@ -0,0 +1,217 @@
+}
+
+func (s *ScriptHistoryFixtureTestSuite) TearDownTest() {
+	s.given.WhatsOnChain().Transport().Reset()
```
when you're creating the testabilities and WoC service in each and every test, then the restet is not needed, because transport is recreated also in each test

jakubmkowalski replied:
removed

dorzepowski commented on pkg/services/internal/whatsonchain/script_history.go:
```diff
@@ -0,0 +1,144 @@
+		SetResult(&history)
+
+	if opts != nil {
+		if opts.Order != nil {
```
No I'm pretty sure that it's not good idea to have those opts,
because we're making 2 calls but those params are only for the confirmed history. So the result could be misleading, because you can request limit 1 and you could get 1 result from here and 1000 from unconfirmed.

jakubmkowalski replied:
opts removed

dorzepowski commented on pkg/services/internal/whatsonchain/script_history.go:
```diff
@@ -0,0 +1,144 @@
+		}
+	}
+
+	return &wdk.ScriptHistoryResult{
```
there is no need to return whole ScriptHistoryResult,
you could simply return the historyItems slice from this method

jakubmkowalski replied:
now we return table of history items

dorzepowski commented on pkg/services/internal/whatsonchain/script_history.go:
```diff
@@ -0,0 +1,144 @@
+		}
+	}
+
+	return &wdk.ScriptHistoryResult{
```
there is no need to return whole ScriptHistoryResult,
you could simply return the historyItems slice from this method

jakubmkowalski replied:
now it returns slice of historyItems

dorzepowski commented on pkg/wdk/services_script_hash_history.go:
```diff
@@ -0,0 +1,108 @@
+}
+
+// ScriptHashHistoryResponse represents the response from the script history endpoints
+type ScriptHashHistoryResponse struct {
```
this doens't need to be here, it rather belongs to WhatsOnChain service
So place it rather in `pkg/services/internal/whatsonchain/internal/dto`

jakubmkowalski replied:
moved

dorzepowski commented on pkg/services/internal/whatsonchain/script_history.go:
```diff
@@ -0,0 +1,144 @@
+			TxHash: item.TxID,
+			Height: item.Height,
+		}
+	}
```
You could create function:
```go
func toScriptHistoryItem(item wdk.ScriptHashHistoryItem) wdk.ScriptHistoryItem {
	return wdk.ScriptHistoryItem{
		TxHash: item.TxID,
		Height: item.Height,
	}
}
```

and then 

```suggestion
	historyItems := slices.Map(history.Result, toScriptHistoryItem)
```

jakubmkowalski replied:
function created, suggestion implemented

dorzepowski commented on pkg/services/internal/whatsonchain/script_history.go:
```diff
@@ -0,0 +1,144 @@
+			TxHash: item.TxID,
+			Height: item.Height,
+		}
+	}
```
this is the same mapping as for unconfirmed, so if you will create function `toScriptHistoryItem`, then here can be:

```suggestion
	historyItems := slices.Map(history.Result, toScriptHistoryItem)
```

jakubmkowalski replied:
changed to suggestion

dorzepowski commented on pkg/services/internal/whatsonchain/script_history.go:
```diff
@@ -0,0 +1,144 @@
+
+	combinedHistory := make([]wdk.ScriptHistoryItem, 0, len(confirmedHistory.History)+len(unconfirmedHistory.History))
+	combinedHistory = append(combinedHistory, confirmedHistory.History...)
+	combinedHistory = append(combinedHistory, unconfirmedHistory.History...)
```
If you would return iter.Seq from the methods above
(so use `seq.Map(seq.FromSlice(history.Result), toScriptHistoryItem)` instead)

then you could change it here, to:

```suggestion
	combinedHistory := seq.Collect(seq.Concat(confirmedHistory, unconfirmedHistory))
```

jakubmkowalski replied:
changed to suggestion

dorzepowski replied:
I prepared a suggestion how to fully implement it, as I suggested in this ☝️ comment.
https://github.com/bsv-blockchain/go-wallet-toolbox/pull/351

jakubmkowalski replied:
applied

dorzepowski commented on pkg/services/internal/whatsonchain/script_history_test.go:
```diff
@@ -0,0 +1,425 @@
+	suite.Run(t, new(ScriptHistoryTestSuite))
+}
+
+func (s *ScriptHistoryTestSuite) SetupTest() {
```
In other tests for WoC we use testabilities to setup the transport and create the service, and I would prefer to have this approach here too, so when something need to be changed, we have single place for doing that.

jakubmkowalski replied:
now we use testabillities to setup

dorzepowski commented on pkg/services/internal/whatsonchain/script_history_test.go:
```diff
@@ -0,0 +1,425 @@
+}
+
+// Helper methods
+func (s *ScriptHistoryTestSuite) registerConfirmedHistoryMock(scriptHash string, status int, response interface{}) {
```
IMHO it is good candidate to be placed in testabilities in WhatsOnChainFixture

jakubmkowalski replied:
moved to testabilities

dorzepowski commented on pkg/services/internal/whatsonchain/script_history_test.go:
```diff
@@ -0,0 +1,425 @@
+		})
+	}
+
+	confirmedResponse := wdk.ScriptHashHistoryResponse{
```
It would be better to use plain string or map, so in case of changing the response struct, those tests would still validate if json parsing of server response is working as expected

jakubmkowalski replied:
changed to plain string

dorzepowski commented on pkg/services/services_get_script_history_test.go:
```diff
@@ -0,0 +1,217 @@
+	s.given.WhatsOnChain().Transport().Reset()
+}
+
+func (s *ScriptHistoryFixtureTestSuite) TestGetScriptHistory_UsingBuilderPattern() {
```
Emmmmmm I don't understand, what is the idea of this test? From the name It looks like there is some builder for GetScriptHistory

jakubmkowalski replied:
removed

dorzepowski commented on pkg/services/services_get_script_history_test.go:
```diff
@@ -0,0 +1,217 @@
+}
+
+func (s *ScriptHistoryFixtureTestSuite) TestGetScriptHistory_ValidationErrors() {
+	invalidHashes := testservices.InvalidScriptHashes()
```
it looks like those InvalidScriptHashes are rather needed for testing this particular method of woc. I'm suggesting to inline this here.

jakubmkowalski replied:
InvalidScriptHashes removed from fixtures

dorzepowski commented on pkg/services/services_get_script_history_test.go:
```diff
@@ -0,0 +1,217 @@
+
+func (s *ScriptHistoryFixtureTestSuite) TestGetScriptHistory_UsingQueryFixture() {
+	// given
+	scriptHashes := testservices.ValidScriptHashes()
```
it looks like those InvalidScriptHashes are rather needed for testing this particular method of woc. I'm suggesting to put it somwhere here in this file or even strictly in test that is using it.

jakubmkowalski replied:
InvalidScriptHashes removed from fixtures

dorzepowski commented on pkg/services/services_get_script_history_test.go:
```diff
@@ -0,0 +1,217 @@
+
+	s.given.WhatsOnChain().
+		WhenQueryingScriptHistory(scriptHash).
+		WillReturnUnconfirmedHistory(http.StatusOK, unconfirmedResponse)
```
I prefer the test with builder then those with plain response in code. 
And I'm not sure if the same couldn't be tested/asserted in the previous test.

jakubmkowalski replied:
changed to use builder

dorzepowski commented on pkg/services/services_get_script_history_test.go:
```diff
@@ -0,0 +1,217 @@
+		WithScriptHash(scriptHash).
+		WithConfirmedTransactions(5, 800000).
+		WithUnconfirmedTransactions(2).
+		SetupMocks(s.given.WhatsOnChain())
```
Maybe:

```suggestion
	s.given.WhatsOnChain().ScriptHistoryData()
		WithScriptHash(scriptHash).
		WithConfirmedTransactions(5, 800000).
		WithUnconfirmedTransactions(2).
		WillBeReturned()
```

jakubmkowalski replied:
changed

dorzepowski commented on pkg/internal/testabilities/testservices/fixture_woc.go:
```diff
@@ -345,3 +354,270 @@ func computeTxID(rawTx []byte) string {
+	confirmedItems := make([]wdk.ScriptHashHistoryItem, b.confirmedCount)
+	for i := 0; i < b.confirmedCount; i++ {
+		confirmedItems[i] = wdk.ScriptHashHistoryItem{
+			TxID:   fmt.Sprintf("confirmed_tx_%064d", i),
```
maybe you could use this, to generate ID that is looking more like ID
```go
func txIDTemplated(index uint) string {
	return fmt.Sprintf("a%010de1b81dd2c9c0c6cd67f9bdf832e9c2bb12a1d57f30cb6ebbe78d9", index)
}
```

jakubmkowalski replied:
changed

dorzepowski commented on pkg/services/services_get_script_history_test.go:
Those tests should be rather in the interna/whatsonchain package, as it is testing the whatsonchain implementation.

And here should be also tests from the walletservices call, but it should have less corner cases, but rather focus on concatenating of results of different service callers.

jakubmkowalski replied:
tests moved to right place, added correct service tests

jakubmkowalski commented on pkg/services/internal/whatsonchain/script_history.go:
```diff
@@ -0,0 +1,144 @@
+}
+
+// GetScriptHistory retrieves both confirmed and unconfirmed script history.
+func (woc *WhatsOnChain) GetScriptHistory(ctx context.Context, scriptHash string, opts *wdk.GetConfirmedScriptHistoryOpts) (*wdk.ScriptHistoryResult, error) {
```
opts removed

jakubmkowalski committed:
feat: update script hash history implementation according to review comments

jakubmkowalski committed:
Refactor script history tests to use builder pattern and improve error handling

- Removed the ScriptHistoryTestSuite and replaced it with standalone test functions.
- Introduced a builder pattern for setting up test data in the script history tests.
- Simplified test cases for valid responses, validation errors, and API errors.
- Enhanced readability and maintainability of test code by reducing boilerplate.
- Consolidated mock responses for confirmed and unconfirmed script history.

jakubmkowalski committed:
feat: introduce to error path tests fixtures builder pattern

jakubmkowalski committed:
feat: add script hash history retry configuration and implement in service

github-actions[bot] committed:
chore: update generated files

jakubmkowalski committed:
Merge branch 'main' into feat/add_script_hash_history

jakubmkowalski committed:
Merge branch 'feat/add_script_hash_history' of github.com:bsv-blockchain/go-wallet-toolbox into feat/add_script_hash_history

github-actions[bot] committed:
chore: update generated files

jakubmkowalski committed:
Merge branch 'main' into feat/add_script_hash_history

jakubmkowalski committed:
feat: refactor script history HTTP client and update retry configuration in service

jakubmkowalski committed:
Merge branch 'feat/add_script_hash_history' of github.com:bsv-blockchain/go-wallet-toolbox into feat/add_script_hash_history

mgosek-4chain commented on pkg/internal/testabilities/testservices/fixture_woc.go:
```diff
@@ -374,3 +382,268 @@ func (f *wocFixture) WillRespondWithBlockHeaderByHeight(status int, height uint3
+		return resp, nil
+	}
+	url := fmt.Sprintf("https://api.whatsonchain.com/v1/bsv/%s/script/%s/confirmed/history", q.fixture.network, q.scriptHash)
+	q.fixture.transport.RegisterResponder("GET", url, responder)
```
`http.MethodGet`

jakubmkowalski replied:
changed

jakubmkowalski committed:
Merge branch 'main' into feat/add_script_hash_history

jakubmkowalski committed:
refactor: replace string method with http.Method constants in responder registrations

jakubmkowalski committed:
Merge branch 'feat/add_script_hash_history' of github.com:bsv-blockchain/go-wallet-toolbox into feat/add_script_hash_history

jakubmkowalski commented on pkg/services/internal/whatsonchain/service.go:
```diff
@@ -59,17 +63,24 @@ func New(httpClient *resty.Client, logger *slog.Logger, network defs.BSVNetwork,
 		SetLogger(logging.RestyAdapter(logger)).
 		SetDebug(logging.IsDebug(logger))
 
+	clientForScriptHashHistory := client.
```
As you can see I created copy of resty client with retry conditions only for script hash history requests.
I think we should consider unify retry conditions for all requests to WOC. What do you think?

mgosek-4chain replied:
Basically you don't have to do this because it's already set (i.e common settings) in the resty factory. 

Please take a look at `httpx` package and `factory.go` file content.

dorzepowski replied:
Agree: no need for copy here and for additional config

jakubmkowalski replied:
thanks guys, duplicate of resty client removed

dorzepowski commented on pkg/services/internal/whatsonchain/service.go:
```diff
@@ -204,3 +215,11 @@ func (woc *WhatsOnChain) PostBEEF(ctx context.Context, beef *transaction.Beef, t
 }
+
+func (woc *WhatsOnChain) GetScriptHashHistory(ctx context.Context, scriptHash string) (*wdk.ScriptHistoryResult, error) {
+	result, err := woc.GetScriptHistory(ctx, scriptHash)
```
what for is this embedding one method inside another one?

IMHO GetScriptHashHistory should be removed
and GetScriptHistory should be renamed to GetScriptHashHistory

jakubmkowalski replied:
right, this method removed, GetScriptHistory renamed

dorzepowski commented on pkg/services/internal/whatsonchain/testabilities/fixtures.go:
```diff
@@ -33,10 +33,12 @@ func (f *wocServiceFixture) NewWoCService(opts ...func(*whatsonchain.WhatsOnChai
+		BroadcastDelay:                 0,
+		BSVExchangeRate:                defs.BSVExchangeRate{},
+		ScriptHashHistoryRetries:       1,
+		ScriptHashHistoryRetryInterval: 0,
```
no need fo this additional confg

jakubmkowalski replied:
removed

dorzepowski commented on pkg/defs/services.go:
```diff
@@ -23,6 +23,12 @@ const (
+	DefaultScriptHashHistoryRetries = 2
+
+	// DefaultScriptHashHistoryRetryInterval is the interval between retries for fetching script hash history
+	DefaultScriptHashHistoryRetryInterval = 2 * time.Second
```
this is also not needed

jakubmkowalski replied:
removed

dorzepowski commented on pkg/defs/whats_on_chain.go:
```diff
@@ -13,6 +13,9 @@ type WhatsOnChain struct {
 	RootForHeightRetries       int             `mapstructure:"root_for_height_retries"`
+
+	ScriptHashHistoryRetries       int           `mapstructure:"script_hash_history_retries"`
+	ScriptHashHistoryRetryInterval time.Duration `mapstructure:"script_hash_history_retry_interval"`
```
this is also not needed

jakubmkowalski replied:
removed

dorzepowski commented on pkg/internal/testabilities/testservices/fixture_services.go:
```diff
@@ -61,6 +61,8 @@ func givenServicesWithNetwork(t testing.TB, network defs.BSVNetwork) ServicesFix
 	servicesConfig.WhatsOnChain.RootForHeightRetries = 1
 	servicesConfig.WhatsOnChain.RootForHeightRetryInterval = 0
+	servicesConfig.WhatsOnChain.ScriptHashHistoryRetries = 1
+	servicesConfig.WhatsOnChain.ScriptHashHistoryRetryInterval = 0
```
this is also no needed 
(sorry for so many comments - but I want to point out, what for sure should be removed)

jakubmkowalski replied:
removed

dorzepowski commented on infra-config.example.yaml:
```diff
@@ -80,3 +80,5 @@ wallet_services:
         root_for_height_retries: 3
         root_for_height_retry_interval: 1s
+        script_hash_history_retries: 0
+        script_hash_history_retry_interval: 0s
```
this is also not needed

jakubmkowalski replied:
removed

dorzepowski commented on pkg/internal/testabilities/testservices/fixture_woc.go:
```diff
@@ -374,3 +382,268 @@ func (f *wocFixture) WillRespondWithBlockHeaderByHeight(status int, height uint3
+	return b
+}
+
+// New method to build raw JSON strings
```
Unneeded comment

jakubmkowalski replied:
removed

dorzepowski commented on pkg/internal/testabilities/testservices/fixture_woc.go:
```diff
@@ -374,3 +382,268 @@ func (f *wocFixture) WillRespondWithBlockHeaderByHeight(status int, height uint3
+}
+
+// New method to build raw JSON strings
+func (b *scriptHistoryDataBuilder) BuildJSON() (confirmedJSON, unconfirmedJSON string) {
```
should be private method

jakubmkowalski replied:
changed to private

dorzepowski commented on pkg/internal/testabilities/testservices/fixture_woc.go:
```diff
@@ -374,3 +382,268 @@ func (f *wocFixture) WillRespondWithBlockHeaderByHeight(status int, height uint3
+	}
+
+	confirmedItems := make([]string, b.confirmedCount)
+	for i := 0; i < b.confirmedCount; i++ {
```
Please create a private methods
for confirmed succes, unconfirmed success, confirmed error and unconfimed error

And choose which one to call based on existance of error

jakubmkowalski replied:
private methods created

jakubmkowalski commented on pkg/services/internal/whatsonchain/service.go:
```diff
@@ -59,17 +63,24 @@ func New(httpClient *resty.Client, logger *slog.Logger, network defs.BSVNetwork,
 		SetLogger(logging.RestyAdapter(logger)).
 		SetDebug(logging.IsDebug(logger))
 
+	clientForScriptHashHistory := client.
```
copy removed so this is resolved

jakubmkowalski committed:
refactor: remove script hash history configuration and related code

jakubmkowalski committed:
Merge branch 'main' into feat/add_script_hash_history
